backend/src/main/java/ai/verta/modeldb/lineage/LineageDAORdbImpl.java

```diff
@@ -69,166 +293,304 @@ private void validate(List<LineageEntry> inputList, List<LineageEntry> outputLis
+      if (connectionEntity.getId().equals(id)) {
+        return;
+      }
+    }
+    session.save(new ConnectionEntity(id, entityId, connectionType, entityType));
   }
 
-  private Optional<LineageEntity> getExisting(
-      Session session, LineageEntry input, LineageEntry output) {
-    Query query =
-        session.createQuery(
-            "from LineageEntity where inputExternalId = '"
-                + input.getExternalId()
-                + "' and inputType = '"
-                + input.getTypeValue()
-                + "' and outputExternalId = '"
-                + output.getExternalId()
-                + "' and outputType = '"
-                + output.getTypeValue()
-                + "'");
-    return Optional.ofNullable((LineageEntity) query.uniqueResult());
+  private List<LineageEntryBatchResponseSingle> getOutputsByInput(
+      Session session,
+      LineageEntryBatchRequest input,
+      BlobHashInCommitFunction blobHashInCommitFunction,
+      CommitInBlobHashFunction commitInBlobHashFunction)
+      throws ModelDBException {
+    return getInputOrOutput(
+        session, input, CONNECTION_TYPE_OUTPUT, blobHashInCommitFunction, commitInBlobHashFunction);
   }
 
-  private List<LineageEntry> getOutputsByInput(Session session, LineageEntry input) {
-    Query query =
-        session.createQuery(
-            "from LineageEntity where inputExternalId = '"
-                + input.getExternalId()
-                + "' and inputType = '"
-                + input.getTypeValue()
-                + "'");
-    List<LineageEntry> result = new LinkedList<>();
-    for (Object r : query.getResultList()) {
-      LineageEntity lineageEntity = (LineageEntity) r;
-      result.add(
-          LineageEntry.newBuilder()
-              .setTypeValue(lineageEntity.getOutputType())
-              .setExternalId(lineageEntity.getOutputExternalId())
-              .build());
-    }
-    return result;
+  private List<LineageEntryBatchResponseSingle> getInputsByOutput(
+      Session session,
+      LineageEntryBatchRequest output,
+      BlobHashInCommitFunction blobHashInCommitFunction,
+      CommitInBlobHashFunction commitInBlobHashFunction)
+      throws ModelDBException {
+    return getInputOrOutput(
+        session, output, CONNECTION_TYPE_INPUT, blobHashInCommitFunction, commitInBlobHashFunction);
   }
 
-  private List<LineageEntry> getInputsByOutput(Session session, LineageEntry output) {
-    Query query =
-        session.createQuery(
-            "from LineageEntity where outputExternalId = '"
-                + output.getExternalId()
-                + "' and outputType = '"
-                + output.getTypeValue()
-                + "'");
-    List<LineageEntry> result = new LinkedList<>();
-    for (Object r : query.getResultList()) {
-      LineageEntity lineageEntity = (LineageEntity) r;
+  private List<LineageEntryBatchResponseSingle> getInputOrOutput(
+      Session session,
+      LineageEntryBatchRequest sideB,
```

Comment:
What is `sideB`? I'm not sure what it means.

Reply:
Added comments: A -- what we want to receive (can be input or output). B -- what we already have (input or output). So sideB information about entries.